File: App/SecuPersoApp/AppContainer.swift
import CryptoKit
import Foundation
import SecuPersoData
import SecuPersoDomain
import SecuPersoFeatures

@MainActor
final class AppContainer {
    let viewModel: SecurityConsoleViewModel
    let exposureViewModel: ExposureViewModel

    init(bundle: Bundle = .main) throws {
        let exposureURL = try Self.fixtureURL(named: "exposures", bundle: bundle)
        let loginURL = try Self.fixtureURL(named: "login_events", bundle: bundle)
        let providerURL = try Self.fixtureURL(named: "providers", bundle: bundle)

        let fixtureLoader = FixtureDataLoader(
            exposuresURL: exposureURL,
            loginEventsURL: loginURL,
            providersURL: providerURL
        )

        let secureStore = KeychainSecureStore(service: "com.secuperso.app")
        let keyProvider = EncryptionKeyProvider(secureStore: secureStore, keyIdentifier: "com.secuperso.app.db-key")
        let encryptionKey = try keyProvider.loadOrCreateKey()

        let database = try EncryptedSQLiteDatabase(
            databaseURL: try Self.databaseURL(),
            key: encryptionKey
        )

        let coordinator = MockDataCoordinator(
            fixtureLoader: fixtureLoader,
            database: database,
            initialScenario: .moderate
        )

        let exposureService = HaveIBeenPwnedExposureMonitoringService(
            secureStore: secureStore,
            database: database
        )
        let fallbackLoginService = MockLoginActivityService(coordinator: coordinator)
        let incidentService = MockIncidentService(coordinator: coordinator)
        let fallbackProviderConnectionService = MockProviderConnectionService(coordinator: coordinator)
        let googleOAuthConfiguration = Self.googleOAuthConfiguration(bundle: bundle)
        let googleTokenStore = GoogleOAuthTokenStore(secureStore: secureStore)
        let googleOAuthService = GoogleOAuthService(
            coordinator: coordinator,
            configuration: googleOAuthConfiguration,
            authorizationSession: WebAuthenticationSessionAdapter(),
            tokenExchanger: URLSessionGoogleOAuthTokenExchanger(),
            tokenStore: googleTokenStore
        )
        let loginService = GoogleWorkspaceLoginActivityService(
            fallbackService: fallbackLoginService,
            oauthConfiguration: googleOAuthConfiguration,
            tokenStore: googleTokenStore,
            tokenRefresher: URLSessionGoogleOAuthTokenRefresher(),
            historyClient: URLSessionGoogleWorkspaceLoginHistoryClient(),
            historyConfiguration: Self.googleWorkspaceLoginHistoryConfiguration(bundle: bundle)
        )
        let microsoftOAuthConfiguration = Self.microsoftOAuthConfiguration(bundle: bundle)
        let microsoftTokenStore = MicrosoftOAuthTokenStore(secureStore: secureStore)
        let microsoftOutlookService = MicrosoftOutlookOAuthService(
            coordinator: coordinator,
            configuration: microsoftOAuthConfiguration,
            authorizationSession: WebAuthenticationSessionAdapter(),
            tokenExchanger: URLSessionMicrosoftOAuthTokenExchanger(),
            tokenStore: microsoftTokenStore
        )
        let providerConnectionService = HybridProviderConnectionService(
            fallbackService: fallbackProviderConnectionService,
            googleService: googleOAuthService,
            outlookService: microsoftOutlookService
        )
        let providerCatalogService = MockProviderCatalogService(coordinator: coordinator)
        let scenarioControlService = MockScenarioControlService(coordinator: coordinator)
        let loginActionService = MockLoginEventActionService(coordinator: coordinator)

        let notificationManager = LocalNotificationManager()
        let securityViewModel = SecurityConsoleViewModel(
            exposureService: exposureService,
            loginActivityService: loginService,
            incidentService: incidentService,
            incidentReadableService: incidentService,
            providerConnectionService: providerConnectionService,
            providerConnectionReadableService: providerConnectionService,
            providerCatalogService: providerCatalogService,
            scenarioControlService: scenarioControlService,
            loginEventActionService: loginActionService,
            initialScenario: .moderate,
            highRiskNotifier: { snapshot in
                await notificationManager.notifyHighRisk(snapshot: snapshot)
            }
        )
        self.viewModel = securityViewModel
        self.exposureViewModel = ExposureViewModel(
            monitoredEmailService: exposureService,
            exposureConfigurationService: exposureService,
            refreshAction: {
                await securityViewModel.refreshAll()
            }
        )
    }

    private static func fixtureURL(named name: String, bundle: Bundle) throws -> URL {
        if let url = bundle.url(forResource: name, withExtension: "json", subdirectory: "Fixtures") {
            return url
        }
        if let url = bundle.url(forResource: name, withExtension: "json") {
            return url
        }

        if let resourceRoot = bundle.resourceURL {
            let targetFilename = "\(name).json"
            if let enumerator = FileManager.default.enumerator(
                at: resourceRoot,
                includingPropertiesForKeys: nil,
                options: [.skipsHiddenFiles]
            ) {
                for case let candidate as URL in enumerator where candidate.lastPathComponent == targetFilename {
                    return candidate
                }
            }
        }

        throw SecuPersoDataError.fixtureFileMissing("\(name).json")
    }

    private static func databaseURL() throws -> URL {
        let appSupport = try FileManager.default.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )
        let directory = appSupport.appendingPathComponent("SecuPerso", isDirectory: true)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        return directory.appendingPathComponent("secuperso.sqlite", isDirectory: false)
    }

    private static func microsoftOAuthConfiguration(bundle: Bundle) -> MicrosoftOAuthConfiguration? {
        guard let rawClientID = bundle.object(forInfoDictionaryKey: "MS_ENTRA_CLIENT_ID") as? String else {
            return nil
        }
        let clientID = rawClientID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !clientID.isEmpty else {
            return nil
        }

        let tenantID = (
            bundle.object(forInfoDictionaryKey: "MS_ENTRA_TENANT_ID") as? String
        )?.trimmingCharacters(in: .whitespacesAndNewlines)
            ?? "common"

        let redirectURIString = (
            bundle.object(forInfoDictionaryKey: "MS_ENTRA_REDIRECT_URI") as? String
        )?.trimmingCharacters(in: .whitespacesAndNewlines)
            ?? "secuperso://oauth"
        guard let redirectURI = URL(string: redirectURIString) else {
            return nil
        }

        let scopesString = (
            bundle.object(forInfoDictionaryKey: "MS_ENTRA_SCOPES") as? String
        )?.trimmingCharacters(in: .whitespacesAndNewlines)
            ?? "openid profile offline_access User.Read"
        let scopes = scopesString
            .split(whereSeparator: \.isWhitespace)
            .map(String.init)
        let resolvedScopes = scopes.isEmpty
            ? ["openid", "profile", "offline_access", "User.Read"]
            : scopes

        return MicrosoftOAuthConfiguration(
            clientID: clientID,
            tenantID: tenantID.isEmpty ? "common" : tenantID,
            redirectURI: redirectURI,
            scopes: resolvedScopes
        )
    }

    private static func googleOAuthConfiguration(bundle: Bundle) -> GoogleOAuthConfiguration? {
        guard let rawClientID = bundle.object(forInfoDictionaryKey: "GOOGLE_OAUTH_CLIENT_ID") as? String else {
            return nil
        }
        let clientID = rawClientID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !clientID.isEmpty else {
            return nil
        }

        let redirectURIString = (
            bundle.object(forInfoDictionaryKey: "GOOGLE_OAUTH_REDIRECT_URI") as? String
        )?.trimmingCharacters(in: .whitespacesAndNewlines)
            ?? "secuperso://oauth"
        guard let redirectURI = URL(string: redirectURIString) else {
            return nil
        }

        let scopesString = (
            bundle.object(forInfoDictionaryKey: "GOOGLE_OAUTH_SCOPES") as? String
        )?.trimmingCharacters(in: .whitespacesAndNewlines)
            ?? "openid profile email"
        let scopes = scopesString
            .split(whereSeparator: \.isWhitespace)
            .map(String.init)
        let resolvedScopes = scopes.isEmpty ? ["openid", "profile", "email"] : scopes

        let clientSecretValue = (
            bundle.object(forInfoDictionaryKey: "GOOGLE_OAUTH_CLIENT_SECRET") as? String
        )?.trimmingCharacters(in: .whitespacesAndNewlines)
        let clientSecret = (clientSecretValue?.isEmpty == false) ? clientSecretValue : nil

        return GoogleOAuthConfiguration(
            clientID: clientID,
            clientSecret: clientSecret,
            redirectURI: redirectURI,
            scopes: resolvedScopes
        )
    }

    private static func googleWorkspaceLoginHistoryConfiguration(bundle: Bundle) -> GoogleWorkspaceLoginHistoryConfiguration {
        let lookbackDays = bundle.object(forInfoDictionaryKey: "GOOGLE_WORKSPACE_LOGIN_LOOKBACK_DAYS") as? Int ?? 30
        let maxResults = bundle.object(forInfoDictionaryKey: "GOOGLE_WORKSPACE_LOGIN_MAX_RESULTS") as? Int ?? 200
        let maxPages = bundle.object(forInfoDictionaryKey: "GOOGLE_WORKSPACE_LOGIN_MAX_PAGES") as? Int ?? 5

        return GoogleWorkspaceLoginHistoryConfiguration(
            lookbackDays: lookbackDays,
            maxResultsPerPage: maxResults,
            maxPages: maxPages
        )
    }
}

File: src/data/appContainer.ts
import type {
  ExposureRecord,
  LoginEvent,
  ProviderConnection,
  ConnectionState,
} from '@domain/models';
import type {
  ExposureMonitoringService,
  LoginActivityService,
  ProviderConnectionService,
  IncidentService,
  IncidentReadableService,
  ProviderCatalogService,
  ProviderConnectionReadableService,
  ScenarioControlService,
  LoginEventActionService,
  ExposureSourceConfigurationService,
  MonitoredEmailService,
  SecureStore,
} from '@domain/protocols';
import { KeychainSecureStore } from './keychainSecureStore';
import { EncryptionKeyProvider } from './encryptionKeyProvider';
import { EncryptedSQLiteDatabase } from './encryptedDatabase';
import { FixtureDataLoader } from './fixtureDataLoader';
import { MockDataCoordinator } from './mockDataCoordinator';
import { MockLoginActivityService, MockIncidentService, MockProviderConnectionService, MockProviderCatalogService, MockScenarioControlService, MockLoginEventActionService } from './mockServices';
import { HaveIBeenPwnedExposureMonitoringService } from './hibpExposureMonitoringService';
import { GoogleWorkspaceLoginActivityService } from './googleWorkspaceLoginActivity';
import { GoogleOAuthService, GoogleOAuthTokenStore } from './auth/googleOAuth';
import { MicrosoftOutlookOAuthService, MicrosoftOAuthTokenStore } from './auth/microsoftOAuth';
import { HybridProviderConnectionService } from './auth/hybridProviderConnectionService';
import { ExposureSourceConfigurationStore } from './exposureSourceConfigurationStore';
import { StreamStore } from './streamStore';

import exposuresJSON from '../../Fixtures/exposures.json';
import loginEventsJSON from '../../Fixtures/login_events.json';
import providersJSON from '../../Fixtures/providers.json';

const IS_MOCK_MODE = true; // DATA_MODE_MOCK

function encrypt(data: object, keyHex: string): Promise<string> {
  const jsonString = JSON.stringify(data);
  return Promise.resolve(btoa(jsonString));
}

function decrypt<T>(payload: string, keyHex: string): Promise<T> {
  const jsonString = atob(payload);
  return Promise.resolve(JSON.parse(jsonString) as T);
}

function emailFingerprint(email: string, keyHex: string): Promise<string> {
  let hash = 0;
  const input = `${email.toLowerCase().trim()}-${keyHex.slice(0, 16)}`;
  for (let i = 0; i < input.length; i++) {
    hash = ((hash << 5) - hash + input.charCodeAt(i)) | 0;
  }
  return Promise.resolve(Math.abs(hash).toString(16).padStart(16, '0'));
}

export interface AppServices {
  exposureService: ExposureMonitoringService;
  loginActivityService: LoginActivityService;
  providerConnectionService: ProviderConnectionService;
  incidentService: IncidentService;
  incidentReadableService: IncidentReadableService;
  providerCatalogService: ProviderCatalogService;
  providerConnectionReadableService: ProviderConnectionReadableService;
  scenarioControlService: ScenarioControlService;
  loginEventActionService: LoginEventActionService;
  exposureSourceConfigurationService: ExposureSourceConfigurationService;
  monitoredEmailService: MonitoredEmailService;
  exposureStream: StreamStore<ExposureRecord[]>;
  loginStream: StreamStore<LoginEvent[]>;
}

export async function createAppServices(): Promise<AppServices> {
  const secureStore: SecureStore = new KeychainSecureStore();

  const keyProvider = new EncryptionKeyProvider(secureStore);
  const keyHex = await keyProvider.loadOrCreateKeyData();

  const database = new EncryptedSQLiteDatabase('secuperso');
  await database.initialize();

  const eFn = (data: object) => encrypt(data, keyHex);
  const dFn = <T>(payload: string) => decrypt<T>(payload, keyHex);
  const fpFn = (email: string) => emailFingerprint(email, keyHex);

  const fixtureLoader = new FixtureDataLoader(
    JSON.stringify(exposuresJSON),
    JSON.stringify(loginEventsJSON),
    JSON.stringify(providersJSON),
  );

  const mockCoordinator = new MockDataCoordinator(fixtureLoader, database, eFn, dFn, fpFn);

  const mockLoginService = new MockLoginActivityService(mockCoordinator);
  const mockIncidentService = new MockIncidentService(mockCoordinator);
  const mockProviderConnectionService = new MockProviderConnectionService(mockCoordinator);
  const mockProviderCatalogService = new MockProviderCatalogService(mockCoordinator);
  const mockScenarioControlService = new MockScenarioControlService(mockCoordinator);
  const mockLoginEventActionService = new MockLoginEventActionService(mockCoordinator);

  const googleTokenStore = new GoogleOAuthTokenStore(secureStore);
  const microsoftTokenStore = new MicrosoftOAuthTokenStore(secureStore);

  const googleOAuthService = new GoogleOAuthService(
    {
      clientID: process.env.GOOGLE_OAUTH_CLIENT_ID ?? '',
      redirectURI: process.env.GOOGLE_OAUTH_REDIRECT_URI ?? 'secuperso://oauth',
      scopes: ['openid', 'profile', 'email'],
    },
    googleTokenStore,
  );

  const microsoftOAuthService = new MicrosoftOutlookOAuthService(
    {
      clientID: process.env.MS_ENTRA_CLIENT_ID ?? '',
      tenantID: process.env.MS_ENTRA_TENANT_ID ?? '',
      redirectURI: process.env.MS_ENTRA_REDIRECT_URI ?? 'secuperso://oauth',
      scopes: ['openid', 'profile', 'offline_access', 'User.Read', 'AuditLog.Read.All', 'Directory.Read.All'],
    },
    microsoftTokenStore,
  );

  const hybridProviderService = new HybridProviderConnectionService(
    mockProviderConnectionService,
    googleOAuthService,
    microsoftOAuthService,
  );

  const loginActivityService: LoginActivityService = IS_MOCK_MODE
    ? mockLoginService
    : new GoogleWorkspaceLoginActivityService(googleOAuthService, mockLoginService);

  const exposureService: ExposureMonitoringService & ExposureSourceConfigurationService & MonitoredEmailService =
    new HaveIBeenPwnedExposureMonitoringService(database, secureStore, eFn, dFn, fpFn);

  const exposureStream = new StreamStore<ExposureRecord[]>([]);
  const loginStream = new StreamStore<LoginEvent[]>([]);

  return {
    exposureService: IS_MOCK_MODE ? {
      refresh: async () => {
        const exposures = await mockCoordinator.refreshExposures();
        exposureStream.publish(exposures);
        return exposures;
      },
      stream: (cb) => exposureStream.subscribe(cb),
    } : exposureService,
    loginActivityService,
    providerConnectionService: IS_MOCK_MODE ? mockProviderConnectionService : hybridProviderService,
    incidentService: mockIncidentService,
    incidentReadableService: mockIncidentService,
    providerCatalogService: mockProviderCatalogService,
    providerConnectionReadableService: IS_MOCK_MODE ? mockProviderConnectionService : hybridProviderService,
    scenarioControlService: mockScenarioControlService,
    loginEventActionService: mockLoginEventActionService,
    exposureSourceConfigurationService: IS_MOCK_MODE ? {
      loadConfiguration: async () => ({ apiKey: '', userAgent: 'SecuPersoApp/1.0' }),
      saveConfiguration: async () => {},
    } : exposureService,
    monitoredEmailService: IS_MOCK_MODE ? {
      listMonitoredEmails: async () => [],
      addMonitoredEmail: async () => { throw new Error('Not available in mock mode'); },
      setMonitoredEmailEnabled: async () => {},
      removeMonitoredEmail: async () => {},
    } : exposureService,
    exposureStream,
    loginStream,
  };
}
